[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51120 [QNNPACK Sparsity] Add aarch64 kernel of 8x1 sparsity
* #51119 [QNNPACK, Sparsity] Add 8x1 block sparse kernels for aarch32.
* #51118 [QNNPACK, Sparsity] Code refactoring to allow for more generic block
* #50591 [QNNPACK, Sparsity] Add dyanmic linear sparse kernel for arm64
* #50590 [QNNPACK, Sparsity] Sparse kernel with 4x8 blocking
* #50589 [QNNPACK, Sparsity] Added prepacking base aarch32 kernels
* **#50588 [QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear**

Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

kimishpatel committed:
[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear

Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm_sparse/8x4c1x4-dq-aarch32-neon.S:
```diff
@@ -0,0 +1,701 @@
+    # Add offset to r1
+    # Sub offset from r1
+    # This we have to do because we dont have a vector load
+    # instruction with pre-indexed register offset
```
These comments refer to the above block, right?

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm_sparse/8x4c1x4-dq-aarch32-neon.S:
```diff
@@ -0,0 +1,701 @@
+    SUBS r11, r11, 2
+    BLO 1f
+
+k_loop:
```
Yeah, 32 bytes probably is too conservative. I am not sure either whats the best. I kept what was there in the original kernels. But good catch. I think you want alignment that gives minimal # of cache lines for best perf. And in that case I would think 64 might be right even though that is also very conservative. Other than that I dont know how these values are calculated. I see 5, 4 and 3 in original kernels.

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]

kimishpatel committed:
Update on "[QNNPACK, Sparsity] ARMV7, aarch32, kernels for dynamic linear"


Summary:
This diff introduces aarch32 asm kernel for sparse dense gemm.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25925498](https://our.internmc.facebook.com/intern/diff/D25925498)

[ghstack-poisoned]
